static/js/panel.js

```diff
@@ -126,4 +181,131 @@ var Panel = new (class Panel {
+        copy.classList.add("copied");
+        setTimeout(function() {
+          if (!copy.hasAttribute("data-copying")) {
+            copy.classList.remove("copied");
+          }
+        }, 1000);
+      })
+      .finally(function() {
+        copy.removeAttribute("data-copying");
+      });
+  }
+
+  copyMarkdown(type) {
+    const { node, getText } = this.markdown[type];
+    if (node.classList.contains("disabled")) {
+      return;
+    }
+
+    const copy = node.querySelector(".copy");
+    const url = this.permalinkNode ? this.permalinkNode.href : document.location.href;
+    const filename = new URL(url).pathname.match(/\/([^\/]+)$/)[1];
+    const text = getText(url, filename);;
+
+    this.copyText(copy, text);
+  }
+
+  getLanguageFor(filename) {
+    filename = filename.replace(/\.in$/, "");
+
+    const langs = {
+      // suffix => language
+      ".c": "c",
+      ".cc": "cpp",
+      ".configure": "python",
+      ".cpp": "cpp",
+      ".css": "css",
+      ".diff": "diff",
+      ".h": "cpp",
+      ".headers": "http",
+      ".hh": "cpp",
+      ".hpp": "cpp",
+      ".htm": "html",
+      ".html": "html",
+      ".java": "java",
+      ".js": "js",
+      ".jsm": "js",
+      ".json": "json",
+      ".jsx": "js",
+      ".m": "c",
+      ".mathml": "mathml",
+      ".md": "md",
+      ".mjs": "js",
+      ".mm": "cpp",
+      ".mozbuild": "py",
+      ".patch": "diff",
+      ".pl": "perl",
+      ".py": "python",
+      ".rs": "rust",
+      ".rst": "rest",
+      ".scss": "css",
+      ".sjs": "js",
+      ".svg": "xml",
+      ".toml": "toml",
+      ".ts": "js",
+      ".xht": "xhtml",
+      ".xhtml": "xhtml",
+      ".xml": "xml",
+      ".xul": "xul",
+      ".yaml": "yaml",
+      ".yml": "yaml",
+      "^headers^": "http",
+      "moz.build": "py",
+    };
+
+    for (const [suffix, lang] of Object.entries(langs)) {
+      if (filename.endsWith(suffix)) {
+        return lang;
+      }
+    }
+
+    return "";
+  }
+
+  formatSelectedLines() {
+    const texts = [];
+    let lastLine = -1;
+    for (const line of this.selectedLines) {
+      if (lastLine !== -1 && lastLine != line - 1) {
+        texts.push("...");
+      }
+
+      const lineElem = document.getElementById(`line-${line}`).querySelector(".source-line");
+      texts.push(lineElem.textContent.replace(/\n/g, ""));
+
+      lastLine = line;
+    }
+    return texts;
+  }
+
+  updateMarkdownState() {
+    for (const [_, { node, isEnabled }] of Object.entries(this.markdown)) {
+      if (isEnabled()) {
+        node.classList.remove("disabled");
+        node.removeAttribute("aria-disabled");
+        node.querySelector(".copy").disabled = false;
+      } else {
+        node.classList.add("disabled");
+        node.setAttribute("aria-disabled", "true");
+        node.querySelector(".copy").disabled = true;
+      }
+    }
+  }
+
+  onSelectedLineChanged(selectedLines) {
+    this.selectedLines = [...selectedLines].sort((a, b) => a - b);
```

Comment:
It'd be great to avoid this if not copying, is there any reason you need `this.selectedLines` rather than just poking at `Highlight.selectedLines` when needed?

Reply:
Can't you use `self.Highlight && Highlight.selectedLines.length > 0`?

Reply:
if `isEnabled` returns false until `Highlight` global variable is set,
then buttons aren't enabled when loading URL with line number selected.

Reply:
But same happens now right, `Highlight` is what notifies of the selected lines, maybe something like, instead of notifying from the constructor, just notify after `Highlight` definition?

Reply:
Currently the buttons are enabled when loading URL with line numbers selected, because the necessary information is copied from the `Highlight` instance to `Panel` instance, without requiring `Highlight` global variable.

But yeah, if adding extra method call after `Highlight` global variable definition is acceptable, that works for me.